### ISSUE TITLE
YOLO export not exporting fp16 for torchscript

### ISSUE THREAD

**Issue by phillies** [ultralytics/ultralytics#22306]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Export

### Bug

When exporting YOLO as torchscript, the `half` seems to be broken. The size of the exported model is almost identical when exporting YOLO with or without half precision:
```python
from ultralytics import YOLO
import shutil

model = YOLO("yolo11m-seg.pt", task="segment").cuda()

model.export(format="torchscript", device="cuda:0", half=True)
shutil.move("yolo11m-seg.torchscript", "yolo11m-seg_fp16.torchscript")

model.export(format="torchscript", device="cuda:0", half=False)
shutil.move("yolo11m-seg.torchscript", "yolo11m-seg_fp32.torchscript")

model.export(format="onnx", device="cuda:0", half=True)
shutil.move("yolo11m-seg.onnx", "yolo11m-seg_fp16.onnx")
```
This results in the following files:
```bash
$ ll yolo11m*
-rw-r--r-- 1 ubuntu ubuntu 45400152 Oct  1  2024 yolo11m-seg.pt
-rw-r--r-- 1 ubuntu ubuntu 45029455 Oct  7 14:16 yolo11m-seg_fp16.onnx
-rw-r--r-- 1 ubuntu ubuntu 90309737 Oct  7 14:16 yolo11m-seg_fp16.torchscript
-rw-r--r-- 1 ubuntu ubuntu 90327083 Oct  7 14:16 yolo11m-seg_fp32.torchscript
```
The size of the torchscript fp16 and fp32 exports is almost identical. The onnx fp16 export shows the expected size.

I drilled it down to this line:
https://github.com/ultralytics/ultralytics/blob/29592159033715ecac2aa148599e8ac28f637b57/ultralytics/engine/exporter.py#L466
Where the half precision is only applied when it's an onnx export. 

According to the documentation, FP16 should be supported by torchscript, onnx, engine/tensorrt, openvino, and coreml (at least, did not check all formats).

I patched it locally and the export now works as expected:
```bash
-rw-r--r-- 1 ubuntu ubuntu 45465193 Oct  7 14:20 yolo11m-seg_fp16_patched.torchscript
```

Is my understanding wrong, or should this be a condition like `self.args.half and (onnx or jit or engine or ...) and self.device.type != "cpu"`?

Happy to contribute a fix if this is not intended behavior.

### Environment

```
Ultralytics 8.3.206 🚀 Python-3.12.11 torch-2.8.0+cu128 CUDA:0 (NVIDIA GeForce RTX 2080 Ti, 10820MiB)
Setup complete ✅ (48 CPUs, 62.6 GB RAM, 787.3/914.8 GB disk)

OS                     Linux-5.15.0-143-generic-x86_64-with-glibc2.39
Environment            Docker
Python                 3.12.11
Install                pip
Path                   /xxx/.venv/lib/python3.12/site-packages/ultralytics
RAM                    62.64 GB
Disk                   787.3/914.8 GB
CPU                    AMD Ryzen Threadripper 3960X 24-Core Processor
CPU count              48
GPU                    NVIDIA GeForce RTX 2080 Ti, 10820MiB
GPU count              2
CUDA                   12.8

numpy                  ✅ 1.26.4>=1.23.0
matplotlib             ✅ 3.10.6>=3.3.0
opencv-python          ✅ 4.11.0.86>=4.6.0
pillow                 ✅ 11.0.0>=7.1.2
pyyaml                 ✅ 6.0.2>=5.3.1
requests               ✅ 2.32.5>=2.23.0
scipy                  ✅ 1.16.2>=1.4.1
torch                  ✅ 2.8.0+cu128>=1.8.0
torch                  ✅ 2.8.0+cu128!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision            ✅ 0.23.0+cu128>=0.9.0
psutil                 ✅ 7.1.0
polars                 ✅ 1.33.1
ultralytics-thop       ✅ 2.0.17>=2.0.0
```

### Minimal Reproducible Example

```python
from ultralytics import YOLO

model = YOLO("yolo11m-seg.pt", task="segment").cuda()

model.export(format="torchscript", device="cuda:0", half=True)
```

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @phillies, thank you for your interest in Ultralytics 🚀! This is an automated response to help speed things up — an Ultralytics engineer will also assist you here shortly.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

Thanks for the detailed report and for providing both the environment info and an MRE 🙌. Since this concerns Export behavior:
- Please share the full export logs (stdout/stderr) for both `half=True` and `half=False` TorchScript runs
- Confirm the exact model dtype right before export, e.g. `print(model.model.dtype)` and after scripting/tracing if applicable
- If possible, attach both TorchScript artifacts or a minimal diff of file metadata
- Let us know if the behavior changes after a clean re-install/upgrade to the latest `ultralytics` and with a fresh cache
- If you’re preparing a PR, please reference the exporter area you mentioned and outline the conditions you propose so we can review quickly

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
Can you open a PR for it?

There was a user report before and I asked the user to open a PR and forgot about it.

https://github.com/ultralytics/ultralytics/issues/17454#issuecomment-3224938826

**Comment by UltralyticsAssistant** [comment]
Good news! A potential fix for this issue has been merged in the pull request Expand FP16 export support for ultralytics 8.3.208. Full details here: [PR #22316 on GitHub](https://github.com/ultralytics/ultralytics/pull/22316) by @glenn-jocher with contributions from @phillies.

Key updates that may resolve this issue:
- FP16 (half) export behavior is now safer and more predictable:
  - `half=True` is applied on GPU for ONNX and TorchScript (JIT).
  - CPU exports automatically disable FP16 and emit a clear warning.
- JIT now aligns with ONNX in FP16 handling.
- Expanded ONNX FP16 export test coverage to validate these cases.
- Clearer docs on where `half=True` is supported, plus friendlier download error messaging.
- Version bumped to 8.3.208.

How to test the fix:
- Upgrade to the latest version with:
  ```
  pip install -U ultralytics>=8.3.208
  ```
- Or install the latest from the main branch:
  ```
  pip install git+https://github.com/ultralytics/ultralytics.git@main
  ```
- Alternatively, wait for the next official PyPI release if not yet available in your environment.

Quick example to verify:
```python
from ultralytics import YOLO

model = YOLO("yolo11n.pt")

# GPU: FP16 ONNX export (supported)
model.export(format="onnx", half=True, device=0)

# CPU: FP16 will auto-disable with a warning (expected)
model.export(format="onnx", half=True, device="cpu")
```

Feedback on whether this resolves the issue would be greatly appreciated. Thanks for reporting this, and feel free to share more details if anything persists. 🙏
